internal/taskexec/manager.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2025 The A2A Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package taskexec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/a2aproject/a2a-go/a2a"
+	"github.com/a2aproject/a2a-go/a2asrv/eventqueue"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	// ErrExecutionInProgress is returned when a caller attempts to start an execution for
+	// a Task concurrently with another execution.
+	ErrExecutionInProgress = errors.New("task execution is already in progress")
+	// ErrCancelationInProgress is returned when a caller attempts to start an execution for
+	// a Task concurrently with its cancelation.
+	ErrCancelationInProgress = errors.New("task cancelation is in progress")
+)
+
+// Manager provides an API for executing and canceling tasks in a way that ensures
+// concurrent calls don't interfere with one another in unexpected ways.
+// The following guarantees are provided:
+//   - If a Task is being canceled, a concurrent Execution can't be started.
+//   - If a Task is being canceled, a concurrent cancelation will await the existing cancelation.
+//   - If a Task is being executed, a concurrent cancelation will have the same result as the execution.
+//   - If a Task is being executed, a concurrent execution will be rejected.
+//
+// Both cancelations and executions are started in detached context and run until completion.
+type Manager struct {
+	queueManager eventqueue.Manager
+
+	mu           sync.Mutex
+	executions   map[a2a.TaskID]*Execution
+	cancelations map[a2a.TaskID]*cancelation
+}
+
+// NewManager creates an initialized Manager instance.
+func NewManager(queueManager eventqueue.Manager) *Manager {
+	return &Manager{
+		queueManager: queueManager,
+		executions:   make(map[a2a.TaskID]*Execution),
+		cancelations: make(map[a2a.TaskID]*cancelation),
+	}
+}
+
+// GetExecution can be used to resubscribe to events which are being produced by agentExecution.
+func (m *Manager) GetExecution(taskID a2a.TaskID) (*Execution, bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	execution, ok := m.executions[taskID]
+	return execution, ok
+}
+
+// Execute starts an AgentExecutor in a separate goroutine with a detached context.
+// There can only be a single active execution per TaskID.
+func (m *Manager) Execute(ctx context.Context, tid a2a.TaskID, executor Executor) (*Execution, error) {
```

Comment:
(optional) I was about to suggest renaming `GetExecution` to `Execution` (following https://pkg.go.dev/iter#hdr-Naming_Conventions) but I found `Execution` and `Execute` look too similar. 

Another option I can think of is `Execution` for lookup, and `StartExecution`.

Reply:
From the callsite, I think we will assign the manager to a var, so it will often look like

```
p.updateManager.Get(tid)
```

Since this is an internal api, I will stop bikeshedding. :-P btw this question applies more to the other usages of names like `Get` or `GetOrCreate`  in exported interfaces and types.